keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,660 @@
+be admitted. If `FailPolicy.Fail` is used, a webhook outage can result in a
+localized or widespread Kubernetes control plane outage depending on which
+objects the webhook is configured to intercept.
+
+### Goals
+
+- Make CRDs more self-contained and declarative
+- Simplify CRD development
+- Simplify CRD operations
+
+### Non-Goals
+
+- Support for validation formats, immutability or unions. These are all valuable improvements
+  but can be addressed orthogonally in separate KEPs.
+- Eliminate the need for webhooks entirely. Webhooks will still be needed for
+  some use cases. For example, if a validation check requires making a network
+  request to some other system, it will still need to be implemented in a webhook.
+- Support all validation done on native Kubernetes types. For example, CRD structural schemas has
+  complex validation rules that we CEL cannot support due to the lask of support for arbitrarily
+  deeply nested terms (CEL cannot support recursive data types).
+- Change how Kubernetes built-in types are validated, defaulted and converted.
+
+## Proposal
+
+An inline expression language like [Common Expression Language (CEL)](https://github.com/google/cel-go) would be an
+excellent supplement to the current validation mechanism because it is sufficiently expressive to
+satisfy a large set of remaining uses cases that none of the above can solve.
+For example, cross-field validation use cases can only be solved using
+expressions or webhooks.
+
+`x-kubernetes-validator` extension will be added to CRD structural schemas to allow CEL validation expressions.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+...
+  schema:
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          x-kubernetes-validator: 
+            - rule: "minReplicas <= maxReplicas"
+              message: "minReplicas cannot be larger than maxReplicas"
+          type: object
+          properties:
+            minReplicas:
+              type: integer
+            maxReplicas:
+              type: integer
+```
+
+- Each validator may have multiple validation rules.
+
+- Each validation rule has an optional 'message' field for the error message that
+will be surfaced when the validation rule evaluates to false.
+
+- The validator will be scoped to the location of the `x-kubernetes-validator`
+extension in the schema. In the above example, the validator is scoped to the
+'spec' field.
+
+- For OpenAPIv3 object types, the expression will have direct access to all the
+fields of the object the validator is scoped to.
+  
+- For OpenAPIv3 scalar types (integer, string & boolean), the expression will have
+access to the scalar data element the validator is scoped to.
+
+- For OpenAPIv3 list and map types, the expression will have access to the data
+element of the list or map.
+  
+- For immutability use case, validator will have access to the existing version of the object.
```

Comment:
on reflection, it's unclear what would be provided to the expression for the "old data" in all cases.

Imagine an expression on the item schema for the following list:

```
properties:
  spec:
    type: object
    properties:
      refs:
        type: array
        items:
          type: object
          x-kubernetes-validator: ...
          properties:
            name:
              type: string
```

old object:
```
spec:
  refs:
  - name: a
  - name: b
  - name: c
```

new object:
```
spec:
  refs:
  - name: b
  - name: c
  - name: d
```

I would expect the expression to be evaluated three times, with new data of `{"name":"b"}`, `{"name":"c"}`, `{"name":"d"}`, but it's not clear to me what the old data would be

Reply:
fwiw, https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable mapped out a lot of the possible interactions for immutable validation, might be worth linking there

Reply:
`oldProperty`. I've added a additional small section explaining how collisions with CEL keywords will be dealt with as well (since that was clearly needed).  `oldProperty` will be treated as a keyword from an escaping perspective.

Reply:
> on reflection, it's unclear what would be provided to the expression for the "old data" in all cases.

Options:

1. Don't provide old collection types (maps and lists). If you want old, you need to validate the entire list.
2. Provide old only if the type is granular (associative lists can be granular, plain lists cannot): the old data element is the one with the same map key(s), or null if there is no corresponding entry in the old map.  
3. Some variant of the above except we provide the matching positional element as old for plain lists. (I think we should avoid this entirely)

I'm in favor of #2. There are some details to work out, but I'm pretty sure it can be made sane given that we're able to merge granular types.

Reply:
Okay, I think I have a plan for this. It feels closely related to the merge problem.

I've updated the KEP to say it this way: "old is only available on the mergable collection types". I've spelled out what this means and linked to the server side apply documentation explaining atomic and granular.